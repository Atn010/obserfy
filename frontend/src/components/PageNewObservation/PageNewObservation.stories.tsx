import React, { FC } from "react"
import PageNewObservation from "./PageNewObservation"

export default {
  title: "Core|PageNewObservation",
  component: PageNewObservation,
  parameters: {
    componentSubtitle: "Just a simple PageNewObservation"
  }
}

export const Basic: FC = () => <PageNewObservation />
