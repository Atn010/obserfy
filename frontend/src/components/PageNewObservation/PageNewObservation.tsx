import React, { FC, useEffect, useMemo, useState } from "react"
import differenceInCalendarDays from "date-fns/differenceInCalendarDays"
import { useIntl } from "gatsby-plugin-intl3"
import isToday from "date-fns/isToday"
import Box from "../Box/Box"
import { categories } from "../../categories"
import Select from "../Select/Select"
import Input from "../Input/Input"
import TextArea from "../TextArea/TextArea"
import BackNavigation from "../BackNavigation/BackNavigation"
import { useQueryStudentDetails } from "../../hooks/students/useQueryStudentDetails"
import Flex from "../Flex/Flex"
import Button from "../Button/Button"
import Spacer from "../Spacer/Spacer"
import {
  Observation,
  useQueryStudentObservations,
} from "../../hooks/students/useQueryStudentObservations"
import Typography from "../Typography/Typography"
import { getAnalytics } from "../../analytics"
import Pill from "../Pill/Pill"
import Card from "../Card/Card"

interface Props {
  studentId: string
}
export const PageNewObservation: FC<Props> = ({ studentId }) => {
  const [shortDesc, setShortDesc] = useState("")
  const [longDesc, setLongDesc] = useState("")
  const [categoryId, setCategoryId] = useState(categories[0].id)
  const [student] = useQueryStudentDetails(studentId)
  const [
    observations,
    setObservationOutdated,
  ] = useQueryStudentObservations(studentId)

  const [showSuccess, setShowSuccess] = useState(false)
  useEffect(() => {
    const timeOut = setTimeout(() => {
      if (showSuccess) {
        setShowSuccess(false)
      }
    }, 2000)
    return () => clearTimeout(timeOut)
  }, [showSuccess])

  const relatedObservation = useMemo(
    () =>
      observations
        ?.filter(({ createdDate }) => isToday(Date.parse(createdDate ?? "")))
        ?.sort(
          (a, b) => parseInt(a.categoryId, 10) - parseInt(b.categoryId, 10)
        ),
    [observations]
  )

  async function submitAddObservation(observation: Observation): Promise<void> {
    const baseUrl = "/api/v1"
    const response = await fetch(
      `${baseUrl}/students/${studentId}/observations`,
      {
        credentials: "same-origin",
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(observation),
      }
    )
    if (response.status === 201) {
      setObservationOutdated()
      setShortDesc("")
      setLongDesc("")
      setShowSuccess(true)
      setCategoryId(categories[0].id)
    }
    getAnalytics()?.track("Observation Created", {
      responseStatus: response.status,
      observationId: observation.id,
    })
  }

  return (
    <Box maxWidth="maxWidth.md" m="auto">
      <BackNavigation
        to={`/dashboard/students/details?id=${studentId}`}
        text={student?.name ?? ""}
      />
      <Box m={3}>
        <Select
          mb={3}
          label="Category"
          value={categoryId}
          onChange={e => setCategoryId(e.target.value)}
        >
          {categories.map(({ id, name }) => (
            <option key={id} value={id}>
              {name}
            </option>
          ))}
        </Select>
        <Input
          label="Short Description"
          width="100%"
          placeholder="What have you found?"
          onChange={e => setShortDesc(e.target.value)}
          value={shortDesc}
          mb={3}
        />
        <TextArea
          label="Details"
          width="100%"
          fontSize={2}
          height={150}
          placeholder="Tell us about what you observed"
          onChange={e => setLongDesc(e.target.value)}
          value={longDesc}
        />
      </Box>
      <Flex m={3} alignItems="center">
        <Typography.Body
          color="textPrimary"
          opacity={showSuccess ? 1 : 0}
          sx={{
            transition: "opacity 100ms",
          }}
        >
          Observation added!!
        </Typography.Body>
        <Spacer />
        <Button
          onClick={() => {
            submitAddObservation({ longDesc, shortDesc, categoryId })
          }}
        >
          Save
        </Button>
      </Flex>
      <Box m={3}>
        <Typography.H6 mb={3}>Today&apos;s Observations</Typography.H6>
        {relatedObservation?.map(observation => (
          <RelatedObservation key={observation.id} observation={observation} />
        ))}
      </Box>
    </Box>
  )
}

const RelatedObservation: FC<{ observation: Observation }> = ({
  observation,
}) => {
  const category = categories[parseInt(observation.categoryId, 10)]
  const dateDifference = differenceInCalendarDays(
    Date.parse(observation.createdDate ?? ""),
    Date.now()
  )
  const intl = useIntl()
  let date: string
  if (dateDifference > 2) {
    date = intl.formatRelativeTime(dateDifference, "day", { numeric: "auto" })
  } else {
    date = intl.formatDate(observation.createdDate, {
      month: "short",
      year: "numeric",
      weekday: "short",
      day: "numeric",
    })
  }

  return (
    <Card mb={3}>
      <Box mx={3} my={2}>
        <Typography.Body mb={2}>{observation.shortDesc}</Typography.Body>
        <Flex mb={2}>
          <Pill
            backgroundColor={category.color}
            text={category.name}
            color={category.onColor}
          />
          <Pill ml={2} text={date} color="text" />
        </Flex>
        <Typography.Body>{observation.longDesc}</Typography.Body>
      </Box>
    </Card>
  )
}

export default PageNewObservation
