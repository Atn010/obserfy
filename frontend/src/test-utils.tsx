/* eslint-disable import/no-extraneous-dependencies */
import React, { FC, ReactElement } from "react"
import { ThemeProvider } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import { IntlProvider } from "gatsby-plugin-intl3"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import Theme from "./gatsby-plugin-theme-ui"
import Box from "./components/Box/Box"
import en from "./intl/en.json"

const AllProviders: FC = ({ children }) => {
  // TODO: Maybe this better be outside
  const history = createHistory(createMemorySource("/"))
  return (
    <LocationProvider history={history}>
      <ThemeProvider theme={Theme}>
        <IntlProvider locale="en" messages={en}>
          <Box fontSize={[16, 20]}>{children}</Box>
        </IntlProvider>
      </ThemeProvider>
    </LocationProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
