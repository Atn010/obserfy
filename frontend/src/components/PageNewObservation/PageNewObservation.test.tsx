import React from "react"
import { render } from "../../test-utils"
import { Basic } from "./PageNewObservation.stories"

describe("PageNewObservation", () => {
  it("should render correctly", () => {
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
