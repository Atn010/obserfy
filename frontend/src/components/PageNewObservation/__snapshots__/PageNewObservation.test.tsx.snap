// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageNewObservation should render correctly 1`] = `
.emotion-23 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width:40em) {
  .emotion-23 {
    font-size: 20px;
  }
}

.emotion-22 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 700px;
  margin: auto;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: -8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  fill: var(--theme-ui-colors-icon,rgba(0,0,0,0.6));
  margin: 16px;
  height: 24px;
  width: 24px;
  margin-right: 4px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  font-size: 16px;
  margin-bottom: 0;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    font-size: 14px;
  }
}

.emotion-15 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin: 16px;
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  as: label;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding-bottom: 4px;
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 8px;
  color: inherit;
  background-color: transparent;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  background-color: var(--theme-ui-colors-surface,#fff);
  font-size: 16px;
  padding: 16px;
}

.emotion-4:hover,
.emotion-4:focus {
  border-color: var(--theme-ui-colors-primary,#33691e);
  outline: none;
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-left: -28px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  pointer-events: none;
  width: 24px;
  height: 24px;
}

.emotion-11 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  as: label;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  margin-bottom: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-10 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-9 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 8px;
  color: inherit;
  background-color: transparent;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  width: auto;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  background-color: var(--theme-ui-colors-surface,#fff);
  font-size: 16px;
  padding: 16px;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  width: 100%;
}

.emotion-9:hover,
.emotion-9:focus {
  border-color: var(--theme-ui-colors-primary,#33691e);
  outline: none;
}

.emotion-9::-webkit-input-placeholder {
  font-style: italic;
}

.emotion-9::-moz-placeholder {
  font-style: italic;
}

.emotion-9:-ms-input-placeholder {
  font-style: italic;
}

.emotion-9::placeholder {
  font-style: italic;
}

.emotion-9:disabled {
  opacity: 0.38;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  color: var(--theme-ui-colors-textDisabled,rgba(0,0,0,0.38));
}

.emotion-14 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  as: label;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-size: 14px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-13 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: block;
  width: 100%;
  padding: 8px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: inherit;
  line-height: inherit;
  border: 1px solid;
  border-radius: 8px;
  color: inherit;
  background-color: transparent;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  border-color: var(--theme-ui-colors-border,rgba(0,0,0,.1));
  background-color: var(--theme-ui-colors-surface,#fff);
  font-size: 14px;
  padding: 16px;
  height: 180px;
  resize: vertical;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  width: 100%;
  font-size: 16px;
  height: 150px;
}

.emotion-13:hover,
.emotion-13:focus {
  border-color: var(--theme-ui-colors-primary,#33691e);
  outline: none;
}

.emotion-13::-webkit-input-placeholder {
  font-style: italic;
  line-height: 2rem;
}

.emotion-13::-moz-placeholder {
  font-style: italic;
  line-height: 2rem;
}

.emotion-13:-ms-input-placeholder {
  font-style: italic;
  line-height: 2rem;
}

.emotion-13::placeholder {
  font-style: italic;
  line-height: 2rem;
}

.emotion-19 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin: 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-16 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  -webkit-transition: opacity 100ms;
  transition: opacity 100ms;
  color: var(--theme-ui-colors-textPrimary,#003d00);
  font-size: 16px;
  opacity: 0;
}

@media screen and (min-width:40em) {
  .emotion-16 {
    font-size: 14px;
  }
}

.emotion-17 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  pointer-events: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-18 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: inline-block;
  text-align: center;
  line-height: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
  color: white;
  background-color: var(--theme-ui-colors-primary,#33691e);
  border: 0;
  border-radius: 4px;
  cursor: pointer;
  text-transform: capitalize;
  box-sizing: border-box;
  font-size: 14px;
  font-weight: bold;
  color: var(--theme-ui-colors-onPrimary,#fff);
  background-color: var(--theme-ui-colors-primary,#33691e);
  border-radius: 8px;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  -webkit-transition: background-color 125ms cubic-bezier(0.0,0.0,0.2,1);
  transition: background-color 125ms cubic-bezier(0.0,0.0,0.2,1);
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
  white-space: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-18:disabled {
  opacity: 0.38;
  cursor: auto;
}

.emotion-18:hover,
.emotion-18:focus {
  background-color: var(--theme-ui-colors-primaryDark,#003d00);
  outline: none;
}

.emotion-18:hover:disabled,
.emotion-18:focus:disabled {
  background-color: var(--theme-ui-colors-primary,#33691e);
}

.emotion-20 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 24px;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 300;
  line-height: heading;
  font-size: 2rem;
  line-height: 3rem;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  margin-bottom: 16px;
}

<div>
  <div
    class="emotion-23"
  >
    <div
      class="emotion-22"
    >
      <a
        href="/dashboard/students/details?id=undefined"
      >
        <div
          class="emotion-2"
        >
          <div
            class="emotion-0"
          />
          <p
            class="emotion-1"
          />
        </div>
      </a>
      <div
        class="emotion-15"
      >
        <label
          class="emotion-7"
        >
          <div
            class="emotion-3"
          >
            Category
          </div>
          <div
            class="emotion-6"
          >
            <select
              aria-label="Category"
              class="emotion-4"
            >
              <option
                value="0"
              >
                Others
              </option>
              <option
                value="1"
              >
                Cultural
              </option>
              <option
                value="2"
              >
                Language
              </option>
              <option
                value="3"
              >
                Mathematics
              </option>
              <option
                value="4"
              >
                Practical Life
              </option>
              <option
                value="5"
              >
                Sensorial
              </option>
            </select>
            <svg
              class="emotion-5"
              fill="currentcolor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </label>
        <label
          class="emotion-11"
        >
          <div
            class="emotion-3"
          >
            Short Description
          </div>
          <div
            class="emotion-10"
          >
            <input
              class="emotion-9"
              placeholder="What have you found?"
              value=""
            />
          </div>
        </label>
        <label
          class="emotion-14"
        >
          <div
            class="emotion-3"
          >
            Details
          </div>
          <textarea
            class="emotion-13"
            placeholder="Tell us about what you observed"
          />
        </label>
      </div>
      <div
        class="emotion-19"
      >
        <p
          class="emotion-16"
        >
          Observation added!!
        </p>
        <div
          class="emotion-17"
        />
        <button
          class="emotion-18"
        >
          Save
        </button>
      </div>
      <div
        class="emotion-15"
      >
        <h6
          class="emotion-20"
        >
          Today's Observations
        </h6>
      </div>
    </div>
  </div>
</div>
`;
